compiler/aarch64/codegen/OMRMachine.cpp

```diff
@@ -79,18 +82,380 @@ TR::RealRegister *OMR::ARM64::Machine::freeBestRegister(TR::Instruction *current
+            self()->cg()->freeSpill(location, dataSize, 0);
+
+            if (!self()->cg()->isFreeSpillListLocked())
+               {
+               spilledRegister->setBackingStorage(NULL);
+               }
+            }
+         else
+            {
+            if (debugObj)
+               self()->cg()->traceRegisterAssignment("\nOOL: reverse spill %s in less dominant path (%d / 3), protect spill slot (%p)\n",
+                                             debugObj->getName(spilledRegister), location->getMaxSpillDepth(), location);
+            }
+         }
+      else if (self()->cg()->isOutOfLineHotPath())
+         {
+         // the spilledRegisterList contains all registers that are spilled before entering
+         // the OOL path (in backwards RA). Post dependencies will be generated using this list.
+         // Any registers reverse spilled before entering OOL should be removed from the spilled list
+         if (debugObj)
+            self()->cg()->traceRegisterAssignment("\nOOL: removing %s from the spilledRegisterList\n", debugObj->getName(spilledRegister));
+         self()->cg()->getSpilledRegisterList()->remove(spilledRegister);
+
+         // Reset maxSpillDepth here so that in the cold path we know to free the spill
+         // and so that the spill is not included in future GC points in the hot path while it is protected
+         location->setMaxSpillDepth(0);
+         if (location->getMaxSpillDepth() == 2)
+            {
+            self()->cg()->freeSpill(location, dataSize, 0);
+            if (!self()->cg()->isFreeSpillListLocked())
+               {
+               spilledRegister->setBackingStorage(NULL);
+               }
+            }
+         else
+            {
+            if (debugObj)
+               self()->cg()->traceRegisterAssignment("\nOOL: reverse spilling %s in less dominant path (%d / 2), protect spill slot (%p)\n",
+                                             debugObj->getName(spilledRegister), location->getMaxSpillDepth(), location);
+            }
+         }
+      else // main line
+         {
+         if (debugObj)
+            self()->cg()->traceRegisterAssignment("\nOOL: removing %s from the spilledRegisterList)\n", debugObj->getName(spilledRegister));
+         self()->cg()->getSpilledRegisterList()->remove(spilledRegister);
+         location->setMaxSpillDepth(0);
+         self()->cg()->freeSpill(location, dataSize, 0);
+
+         if (!self()->cg()->isFreeSpillListLocked())
+            {
+            spilledRegister->setBackingStorage(NULL);
+            }
+         }
+      switch (rk)
+         {
+         case TR_GPR:
+            generateMemSrc1Instruction(self()->cg(), TR::InstOpCode::strimmx, currentNode, tmemref, targetRegister, currentInstruction);
+            break;
+         case TR_FPR:
+             generateMemSrc1Instruction(self()->cg(), TR::InstOpCode::vstrimmd, currentNode, tmemref, targetRegister, currentInstruction);
```

Comment:
Same comment as above on commoning these cases up.

Reply:
Merged the GPR and FPR cases.